### PR TITLE
[Feat]Tensor Model Parallel Support For Inference

### PR DESCRIPTION
**Pull request by LRY89757**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 📝 What does this PR do?

- Support the Tensor Parallel for inference

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/inference/modeling/policy/tp_nopadding_llama.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
+from functools import partial
```
Comment (yuanheng-zhao) [inline colossalai/inference/modeling/policy/tp_nopadding_llama.py:1]: It's unnecessary to split llama modeling into two policies. The purpose of policy is to indicate replaced modules/functions/attrs and to manage models under different parallelism settings, and so that a file `tp_nopadding_llama.py` is sort of weird.

File: colossalai/inference/modeling/models/nopadding_llama.py [hunk 2]
```diff
@@ -480,3 +483,343 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
...
+        sequence_lengths: torch.Tensor,
+        cos_sin: Tuple[torch.Tensor],
+        fd_inter_tensor: FDIntermTensors,
+        is_prompts: bool = True,
+        kv_seq_len: int = 0,
+        output_tensor: torch.Tensor = None,
+        sm_scale: int = None,
+        use_cuda_kernel: bool = True,
+        cu_seqlens: torch.Tensor = None,
+        high_precision: bool = False,
```
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/modeling/models/nopadding_llama.py:700]: Yeah, all layers could be merged into one, I will do that in the final commit

File: colossalai/inference/modeling/policy/tp_nopadding_llama.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/modeling/policy/tp_nopadding_llama.py]: Yeah, you're right, I will merge the files and leave one policy only

File: colossalai/inference/core/engine.py [hunk 3]
```diff
@@ -55,51 +61,167 @@ def __init__(
...
-            self.graph_memory_pool = None  # Set during graph capture.
-            if verbose:
-                self.logger.info("Colossal AI CUDA Graph Capture on")
+            cpt_io = TPChekpoint_io()
+            if_has_index_file, model_index_file = self.has_index_file(model_or_path)
+            assert if_has_index_file, "the model path is invalid"
+            cpt_io.load_model(self.model, model_index_file)
 
-            self.capture_model(self.k_cache, self.v_cache)
+        torch.cuda.set_device(self.device)
```
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/core/engine.py]: yeah, fix it

File: colossalai/inference/core/engine.py [hunk 4]
```diff
@@ -55,16 +60,72 @@ def __init__(
...
+
+        self.verbose = verbose
+        if verbose:
+            self.logger = get_dist_logger(__name__)
+
+        # enable memory history, which will
+        # add tracebacks and event history to snapshots
+        # torch.cuda.memory._record_memory_history()
+        self.init_model(model_or_path, model_policy)
+        # torch.cuda.memory._dump_snapshot(f"my_snapshot_rank_{dist.get_rank()}.pickle")
```
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/core/engine.py]: done

File: colossalai/inference/modeling/models/nopadding_llama.py [hunk 5]
```diff
@@ -231,15 +237,145 @@ def llama_rmsnorm_forward(
...
+        self,
+        config: LlamaConfig,
+        mlp_gproj_w: torch.Tensor = None,
+        mlp_uproj_w: torch.Tensor = None,
+        mlp_dproj: ParallelModule = None,
+        process_group: ProcessGroup = None,
+        *args,
+        **kwargs,
+    ):
+        """A Unified Layer for
```
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/modeling/models/nopadding_llama.py:291]: done

File: colossalai/inference/modeling/models/nopadding_llama.py [hunk 6]
```diff
@@ -231,15 +237,145 @@ def llama_rmsnorm_forward(
         return rms_layernorm(hidden_states, self.weight.data, self.variance_epsilon, norm_output, residual)
 
 
-class NopadLlamaAttention(LlamaAttention):
+class NopadLlamaMLP(ParallelModule):
```
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/modeling/models/nopadding_llama.py:280]: What's your point? To inherit some attrs?

File: colossalai/inference/modeling/policy/nopadding_llama.py [hunk 7]
```diff
@@ -25,33 +25,74 @@ def __init__(self) -> None:
...
-        )
+        # add a new item for casual lm
+        new_item = {
+            LlamaForCausalLM: ModulePolicyDescription(
+                sub_module_replacement=[
+                    SubModuleReplacementDescription(
+                        suffix="lm_head", target_module=Linear1D_Col, kwargs={"gather_output": True}
+                    )
+                ],
+                method_replacement={"forward": partial(llama_causal_lm_forward)},
```
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/modeling/policy/nopadding_llama.py]: done, removed

File: colossalai/inference/modeling/policy/nopadding_llama.py [hunk 8]
```diff
@@ -25,33 +25,74 @@ def __init__(self) -> None:
...
 
         self.shard_config._infer()
 
-        infer_forward = llama_causal_lm_forward
-        method_replacement = {"forward": partial(infer_forward)}
-        self.append_or_create_method_replacement(
-            description=method_replacement, policy=policy, target_key=LlamaForCausalLM
-        )
+        # add a new item for casual lm
+        new_item = {
```
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/modeling/policy/nopadding_llama.py]: done

File: colossalai/inference/modeling/models/nopadding_llama.py [hunk 9]
```diff
@@ -249,44 +385,70 @@ def __init__(
...
+    def from_native_module(
+        module: LlamaAttention, process_group: Union[ProcessGroup, List[ProcessGroup]], *args, **kwargs
+    ) -> ParallelModule:
         """Used for initialize the weight of NopadLlamaAttention by origin LlamaAttention.
 
         Args:
             module (LlamaAttention): The origin LlamaAttention layer.
         """
+
+        # LazyInitContext.materialize(module)
```
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/modeling/models/nopadding_llama.py]: removed

File: colossalai/inference/core/plugin.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+from colossalai.cluster import DistCoordinator
+from colossalai.interface import ModelWrapper
+
+try:
+    from torch.nn.modules.module import _EXTRA_STATE_KEY_SUFFIX
+except ImportError:
+    _EXTRA_STATE_KEY_SUFFIX = "_extra_state"
+
+
+class InferCheckpoint_io(GeneralCheckpointIO):
```
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/core/plugin.py:21]: done, added

File: colossalai/inference/modeling/models/nopadding_llama.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (yuanheng-zhao) [reply colossalai/inference/modeling/models/nopadding_llama.py:280]: Yeah. We'd like to inherit original methods as there **might** be some layer calling methods from the original layer as we're going to support more models. The other thing is to explicitly indicate what the module `NopadLlamaMLP` and `NopadLlamaAttention` are based on.

[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/modeling/models/nopadding_llama.py:280]: Got it, Thanks very much! :)

**Comment by LRY89757** [comment]
For random weight, the unit test:
(we compare the TP1, TP2 and the output of Transformers)
image

For real weight(llama-2-hf), the Q&A is Like this:
image

image

File: colossalai/inference/modeling/models/nopadding_llama.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/modeling/models/nopadding_llama.py]: Done

File: colossalai/inference/modeling/models/nopadding_llama.py [hunk 11]
```diff
@@ -231,15 +237,145 @@ def llama_rmsnorm_forward(
...
+        gate_up_proj_out = torch.bmm(hidden_states, self.gate_up_weight)
+        act_out = inference_ops.silu_and_mul(gate_up_proj_out)
+
+        return self.down_proj(act_out)
+
+    def extra_repr(self) -> str:
+        return f"gate_up_proj MergedLinear1D_Col: in_features={self.gate_up_weight.shape[1]}x2, out_features={self.gate_up_weight.shape[2]}, bias=False"
+
+
+class NopadLlamaAttention(ParallelModule):
```
[the remark this answers is not in the document]
Comment (LRY89757) [reply colossalai/inference/modeling/models/nopadding_llama.py]: Done

**Comment by LRY89757** [comment]
* Before this TP pr, the tpt/lat metric:
image
* After this TP pr, the tpt/lat metric:
image

**Review by yuanheng-zhao — APPROVED** [review]
LGTM
